[WIP] Python3 compatible Kobo

frenzymadness opened the pull request:
Hello.

I've done some work and the part of codebase covered by tests is Python 3 compatible!

This PR is still in WIP state because I want to do my own code review after a day or two of rest. In the meantime, feel free to do a review. I am not planning to change something heavily.

The most changes are related to Kobo core which has also the best test coverage but some simple changes like absolute imports, renamed modules and attributes etc. are fixed in all places.

If you find a bug, the best solution is to send a patch with a new test which tests that it won't happen again. But I know that it is not possible very time. Also, if you find something weird, let me know and please add some example of usage, because I am not a Kobo user and I've met it for the first time so maybe I misunderstood something during porting.

Thank you. Hopefully, we can make Kobo Python 3 compatible together.

frenzymadness commented:
Somebody added some Python 3 incompatible test during my porting. I am gonna fix it ASAP. This won't happen again when tests with Python 3 will be enabled.

frenzymadness commented:
Ok, I am not able to fix this. @rohanpm could you please take a look at this error? It seems that `response.status_code` should be somehow set up to some valid integer value when mocking `kobo.hub.views.render_to_response` but I don't know where and how.

Traceback is
```
Executing tests in test_view_log.py                        
[ FAILED ]
...........E.....
======================================================================
ERROR: test_view_zipped_big_html_context (__main__.TestViewLog)
Verify the context passed into HTML template contains correct values.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_view_log.py", line 252, in test_view_zipped_big_html_context
    self.get_log('zipped_big.log')
  File "test_view_log.py", line 479, in get_log
    return self.client.get(url, data)
  File "/home/lbalhar/Software/kobo/.tox/py36-django18/lib/python3.6/site-packages/django/test/client.py", line 500, in get
    **extra)
  File "/home/lbalhar/Software/kobo/.tox/py36-django18/lib/python3.6/site-packages/django/test/client.py", line 303, in get
    return self.generic('GET', path, secure=secure, **r)
  File "/home/lbalhar/Software/kobo/.tox/py36-django18/lib/python3.6/site-packages/django/test/client.py", line 379, in generic
    return self.request(**r)
  File "/home/lbalhar/Software/kobo/.tox/py36-django18/lib/python3.6/site-packages/django/test/client.py", line 466, in request
    six.reraise(*exc_info)
  File "/home/lbalhar/Software/kobo/.tox/py36-django18/lib/python3.6/site-packages/django/utils/six.py", line 686, in reraise
    raise value
  File "/home/lbalhar/Software/kobo/.tox/py36-django18/lib/python3.6/site-packages/django/core/handlers/base.py", line 230, in get_response
    response = self.apply_response_fixes(request, response)
  File "/home/lbalhar/Software/kobo/.tox/py36-django18/lib/python3.6/site-packages/django/core/handlers/base.py", line 277, in apply_response_fixes
    response = func(request, response)
  File "/home/lbalhar/Software/kobo/.tox/py36-django18/lib/python3.6/site-packages/django/http/utils.py", line 30, in conditional_content_removal
    if 100 <= response.status_code < 200 or response.status_code in (204, 304):
TypeError: '<=' not supported between instances of 'int' and 'MagicMock'

----------------------------------------------------------------------
Ran 17 tests in 4.924s
```

tkdchen commented on kobo/decorators.py:
```diff
@@ -62,10 +62,10 @@ def new_func(*args, **kwargs):
             import kobo.tback
-            date = datetime.datetime.strftime(datetime.datetime.now(), "%F %R:%S")
-            data =  "--- TRACEBACK BEGIN: %s ---\n" % date
+            date = datetime.datetime.strftime(datetime.datetime.now(), "%F %R:%S").encode()
```
Isn't an explicit encoding necessary here? For example, ``utf-8``.

frenzymadness replied:
No, because `utf-8` is the default value for encoding.

rohanpm commented:
I'm working on fixing other tests that I've recently introduced/changed which now fail on python 3.

rohanpm commented:
@frenzymadness https://github.com/rohanpm/kobo/commits/python3-dev as at commit https://github.com/rohanpm/kobo/commit/b9fe3cb79109865efc8f0ebf8a43ecc09ea16c22 should fix the test failures, do you want to have a look and maybe update this pull request?

lubomir commented:
Some work is needed in the `kobo.rpmlib` and `kobo.pkgset` modules. It's currently totally unclear where to expect `str` and where `bytes`. At least `get_header_field` function is hard to use: when asking about arch, it returns `str` for source packages and `bytes` for binary ones.

Changelogs should probably also be decoded (this is tricky as there is no guarantee about what encoding is actually used).

frenzymadness commented:
@rohanpm Great! Thank you! Because I think that all of your changes are reasonable, I merged your branch into mine which resolves tests and conflicts in one.

@lubomir Do you have time to do it? If no, please let me know and I'll try it.

ktdreyer commented:
@frenzymadness thank you very much for doing this!

I've enabled your branch in Travis CI for my application that uses `kobo.conf` and it is working well. https://github.com/red-hat-storage/rhcephcompose/pull/68 . Of course I'm only testing `kobo.conf` :)

frenzymadness commented:
@ktdreyer Thank you for the good news!

frenzymadness commented:
I reviewed all changes and I think that PR is ready. Any other reviews and tests would be very appreciated!

When you find something broken, please let me know and add some use case so I can prepare a test for it.

frenzymadness commented:
I installed Pungi (another Fedora releng tool) which depends on Kobo and run all Pungis tests with Kobo from this PR and everything works. Pungi is Python 2 only.

It's another proof that these changes don't affect backward compatibility.

lubomir commented on kobo/rpmlib.py:
```diff
@@ -69,7 +71,7 @@ def get_rpm_header(file_name, ts=None):
         ts.setVSFlags(rpm._RPMVSF_NOSIGNATURES|rpm._RPMVSF_NODIGESTS)
 
-    if type(file_name) in (str, unicode):
+    if type(file_name) in (str, six.text_type):
```
Should this maybe be changed to `isinstance(file_name, six.string_types)`?

frenzymadness replied:
Yes, you are right. I fixed this in the latest fixup commit. Thanks!

lubomir commented:
With [my Pungi port to Python 3](https://pagure.io/pungi/pull-request/739) and this PR, all Pungi unit tests pass. There are a few issues when running it due to the `str`/`bytes` issues mentioned above.

lubomir commented:
I have looked at the string types in rpmlib and opened https://github.com/frenzymadness/kobo/pull/1 with suggested changes. What do you think?

I have also found one possible issues: `kobo.shortcuts.run` on Python 3 returns `bytes` with this patch. Should the output be decoded? If it's not, then basically all callers will need to be updated to either do the decoding or handle the binary data. So it's easier to do it here, but there's potential danger if someone uses the function to read actual binary non-text data. It's possible, but unlikely.

dmach commented:
@frenzymadness you probably haven't noticed python-3-dev branch which is there for about a year.
Could you include (probably manually backport) some changes from that branch?

frenzymadness commented:
I fixed one issue mentioned by @lubomir and I also merged his PR into this one. Tests are still passing.

Other tests and/or reviews are welcome.

frenzymadness commented:
@dmach Thanks for your note. No, I know `python-3-dev` branch and I really wanted to use it as start point but I decided to not do it and start from scratch because:
- The branch is really old (24 commits ahead, 61 commits behind master) and it is full of conflicts. I tried to merge it to master.
- It contains a lot of backward incompatible changes - mostly usage of new names of modules not present in Python 2.6 and 2.7 (like urllib.parse for example)
- The commits in this branch are full of code style improvements (deleted blank lines, deleted spaces around operators, added spaces in comments etc.) so it is hard to review the changes and find changes related to Python 3. For example this commit is full of changes but only minor of them related to Python 3: https://github.com/release-engineering/kobo/commit/a78739b7fcc8b99042b57ac8e3a305e0f5899704

It would be awesome if you can point us to some commits/changes in `python-3-dev` not covered with this PR. Then I can port it manually.

Thank you one more time. Have a nice day.

rohanpm commented on kobo/worker/__init__.py:
```diff
@@ -2,10 +2,7 @@
 from __future__ import absolute_import
-import os
-
-from .task import *
```
This is causing a crash in one of kobo's own tasks which in turn causes basic top-level imports to fail:

````
python -c 'import kobo.worker'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python2.6/site-packages/kobo/worker/__init__.py", line 10, in <module>
    TaskContainer.register_module(tasks)
  File "/usr/lib/python2.6/site-packages/kobo/plugins.py", line 173, in register_module
    __import__(module.__name__, {}, {}, module_list)
  File "/usr/lib/python2.6/site-packages/kobo/worker/tasks/task_shutdown_worker.py", line 4, in <module>
    from kobo.worker import TaskBase
ImportError: cannot import name TaskBase
````

At the very least this has to be fixed within kobo itself, but it could well break other code outside of kobo as well since it's a source-incompatible change to stop exporting these names here.

frenzymadness replied:
Thank you for your review! I'll add `from .task import *` back.

lubomir commented on kobo/shortcuts.py:
```diff
@@ -289,7 +290,7 @@ def run(cmd, show_cmd=False, stdout=False, logfile=None, can_fail=False, workdir
             command = "COMMAND: %s\n%s\n" % (cmd, "-" * (len(cmd) + 9))
             if stdout:
-                print command,
+                print(command, end=' ')
```
Should this be `end=''` instead? The printed string ends with a newline, so the space is printed at beginning of next line. Running `python -c 'import kobo.shortcuts; kobo.shortcuts.run("echo foo", show_cmd=True, stdout=True)'` shows different output than before migration (on any Python version).

This change is also included in https://github.com/frenzymadness/kobo/pull/2

frenzymadness replied:
Yes, you are right, thank you!

lubomir commented:
I have created another update: https://github.com/frenzymadness/kobo/pull/2 There were some argument combinations in `shortcuts.run` that still crashed on Python 3. Now there should be tests for all combinations of `universal_newlines`, `show_cmd`, `logfile` and `stdout` arguments. That should hopefully cover everything.

rohanpm commented on kobo/hub/models.py:
```diff
@@ -22,9 +25,11 @@
 
 import kobo.django.fields
-from kobo.client.constants import *
+from kobo.client.constants import TASK_STATES, FINISHED_STATES
```
FAILED_STATES is also required here.

frenzymadness replied:
Fixed, thanks.

rohanpm commented on kobo/rpmlib.py:
```diff
@@ -69,7 +71,7 @@ def get_rpm_header(file_name, ts=None):
         ts.setVSFlags(rpm._RPMVSF_NOSIGNATURES|rpm._RPMVSF_NODIGESTS)
 
-    if type(file_name) in (str, unicode):
+    if type(file_name) in (str, six.string_types):
```
I have a regression in my application relating to this change.  Reproducer is like this:

````
python -c "import kobo.rpmlib; kobo.rpmlib.get_rpm_header(u'quux')"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python2.6/site-packages/kobo/rpmlib.py", line 79, in get_rpm_header
    hdr = ts.hdrFromFdno(fo.fileno())
AttributeError: 'unicode' object has no attribute 'fileno'
````

According to docs six.string_types is basestring() in python 2.  It makes sense then, that a unicode string now fails to match this precise matching on type.

Should it instead be written like this?

````
if isinstance(file_name, six.string_types):
````

frenzymadness replied:
Yes, you are right. Your solution is good for both Pythons:
- In Python 2 `six.string_types` contains `basestring()` which is a base class for `str` and `unicode`
- In Python 3 `six.string_types` contains new `str`

Fixed in the latest commit. Thank you very much.

lubomir commented:
For testing I have created a COPR repo with python 3 builds of kobo:
https://copr.fedorainfracloud.org/coprs/lsedlar/py3-kobo/
The spec lives in my fork of fedora dist-git:
https://src.fedoraproject.org/fork/lsedlar/rpms/kobo/commits/py3
The packaging is not completely correct (at least kobo-admin subpackage is definitely broken and upgrade from kobo to python[23]-kobo is not handled at all).

lubomir commented:
The changes here work well with [compose-utils](https://pagure.io/compose-utils/pull-request/58). (Admittedly it only uses a very small chunk of the functionality.)

frenzymadness commented:
@lubomir Thank you for a lot of work. I can help you also with packaging if you need some assistance. Let me know.

Thanks all of you guys for testing and helping deliver this new Kobo!

rohanpm commented:
👍 Tested with our application Pub, and its test suite is now fully passing with this version of kobo.

Would anyone like to suggest an acceptance criteria for merging this?  Even with our testing, I feel there's a moderate risk of regressions since kobo's own test suite is not thorough.  But nevertheless, it's a pretty good time to merge this since there's no other outstanding changes to be released in kobo soon.

ktdreyer commented:
I agree now would be a good time to merge this and fix further issues as they come up.

frenzymadness commented:
Yes, I agree as well. I think that we tested this PR well. Sure, we cannot test everything but we wait forever also doesn't make any sense.

IMHO good approach could be to merge this PR and set up an internal rule that every future bugfix or new feature should come with tests.

frenzymadness commented on kobo/shortcuts.py:
```diff
@@ -272,8 +271,14 @@ def run(cmd, show_cmd=False, stdout=False, logfile=None, can_fail=False, workdir
     if type(cmd) in (list, tuple):
-        import pipes
-        cmd = " ".join((pipes.quote(i) for i in cmd))
+        if six.PY2:
```
Yes, you are right! Fixed. Thank you!

frenzymadness commented:
I am trying to solve the problem with the build. It seems to be related to different six version outside tox's virtualenv and enabled site packages.

frenzymadness commented:
I was right. Travis has six version 1.5.2 installed and tox has to have enabled site packages because rpmlib cannot be installed via pip. I'll try to solve this issue.

frenzymadness commented:
Problem with Travis wasn't in Travis or Tox but it was caused by hardcoded shebang it test runner. I've made tests runner Python 3 compatible and I've changed shebang.

frenzymadness commented:
If anybody interested in running tests of Kobo in Docker container, see https://github.com/release-engineering/kobo/issues/46

tyroneabdy commented:
Aside from maybe some minor improvement in usability, I think this is fine to be merged and regressions be fixed when we run into them, otherwise if we keep this pull request going too long it'll potentially suffer the chance of disinterest.

rohanpm commented:
@frenzymadness you've got one fixup! here.  Just checking whether you want to adjust the history at all before this is merged.

frenzymadness commented:
I've rebased the whole PR to the master and squashed fixup commit. I don't want to adjust commits history more to keep it in the same order as comments in this PR.
